File: elixir/complex-numbers/lib/complex_numbers.ex
defmodule ComplexNumbers do
  @typedoc """
  In this module, complex numbers are represented as a tuple-pair containing the real and
  imaginary parts.
  For example, the real number `1` is `{1, 0}`, the imaginary number `i` is `{0, 1}` and
  the complex number `4+3i` is `{4, 3}'.
  """
  @type complex :: {number, number}

  @doc """
  Return the real part of a complex number
  """
  @spec real(a :: complex) :: number
  def real({r, _i}), do: r

  @doc """
  Return the imaginary part of a complex number
  """
  @spec imaginary(a :: complex) :: number
  def imaginary({_r, i}), do: i

  @doc """
  Multiply two complex numbers, or a real and a complex number
  """
  @spec mul(a :: complex | number, b :: complex | number) :: complex
  def mul({r1, i1}, {r2, i2}), do: {r1 * r2 - i1 * i2, i1 * r2 + r1 * i2}
  def mul({r1, i1}, b), do: {b * r1, b * i1}
  def mul(a, {r2, i2}), do: {a * r2, a * i2}
  def mul(a, b), do: {a * b, 0}

  @doc """
  Add two complex numbers, or a real and a complex number
  """
  @spec add(a :: complex | number, b :: complex | number) :: complex
  def add({r1, i1}, {r2, i2}), do: {r1 + r2, i1 + i2}
  def add(a, {r2, i2}), do: {a + r2, i2}
  def add({r1, i1}, b), do: {r1 + b, i1}
  def add(a, b), do: {a + b, 0}

  @doc """
  Subtract two complex numbers, or a real and a complex number
  """
  @spec sub(a :: complex | number, b :: complex | number) :: complex
  def sub({r1, i1}, {r2, i2}), do: {r1 - r2, i1 - i2}
  def sub(a, {r2, i2}), do: {a - r2, -i2}
  def sub({r1, i1}, b), do: {r1 - b, i1}
  def sub(a, b), do: {a - b, 0}

  @doc """
  Divide two complex numbers, or a real and a complex number
  """
  @spec div(a :: complex | number, b :: complex | number) :: complex
  def div({r1, i1}, {r2, i2}) do
    d = r2 ** 2 + i2 ** 2
    {(r1 * r2 + i1 * i2) / d, (i1 * r2 - r1 * i2) / d}
  end

  def div({r1, i1}, b), do: {r1 / b, i1 / b}
  def div(a, {r2, i2}), do: ComplexNumbers.div({a, 0}, {r2, i2})

  @doc """
  Absolute value of a complex number
  """
  @spec abs(a :: complex) :: number
  def abs({r, i}), do: :math.sqrt(r ** 2 + i ** 2)

  @doc """
  Conjugate of a complex number
  """
  @spec conjugate(a :: complex) :: complex
  def conjugate({r, i}), do: {r, -i}

  @doc """
  Exponential of a complex number
  """
  @spec exp(a :: complex) :: complex
  def exp({r1, i1}) do
    {:math.cos(i1), :math.sin(i1)}
    |> ComplexNumbers.mul(:math.exp(r1))
  end
end

File: go/complex-numbers/complex_numbers.go
// Package complexnumbers provides a type for representing and performing operations on complex numbers.
// This package offers a simple interface for performing common operations on complex numbers.
package complexnumbers

import "math"

// Number represents a complex number with real and imaginary parts.
type Number struct {
	real, imaginary float64
}

// Real returns the real part of the complex number.
func (n Number) Real() float64 {
	return n.real
}

// Imaginary returns the imaginary part of the complex number.
func (n Number) Imaginary() float64 {
	return n.imaginary
}

// Add adds two complex numbers and returns the result.
func (n1 Number) Add(n2 Number) Number {
	return Number{
		n1.real + n2.real,
		n1.imaginary + n2.imaginary,
	}
}

// Subtract subtracts one complex number from another and returns the result.
func (n1 Number) Subtract(n2 Number) Number {
	return Number{
		n1.real - n2.real,
		n1.imaginary - n2.imaginary,
	}
}

// Multiply multiplies the complex number by another complex number and returns the result.
func (n1 Number) Multiply(n2 Number) Number {
	return Number{
		n1.real*n2.real - n1.imaginary*n2.imaginary,
		n1.imaginary*n2.real + n1.real*n2.imaginary,
	}
}

// Times multiplies the complex number by a real number and returns the result.
func (n Number) Times(factor float64) Number {
	return Number{
		n.real * factor,
		n.imaginary * factor,
	}
}

// Divide divides the complex number by another complex number and returns the result.
// It uses the formula for dividing complex numbers.
func (n1 Number) Divide(n2 Number) Number {
	factor := (n2.real*n2.real + n2.imaginary*n2.imaginary)
	realPart := (n1.real*n2.real + n1.imaginary*n2.imaginary) / factor
	imaginaryPart := (n1.imaginary*n2.real - n1.real*n2.imaginary) / factor

	return Number{
		realPart,
		imaginaryPart,
	}
}

// Conjugate returns the complex conjugate of the number.
func (n Number) Conjugate() Number {
	return Number{n.real, -n.imaginary}
}

// Abs returns the absolute value (magnitude) of the complex number.
func (n Number) Abs() float64 {
	return math.Sqrt(n.real*n.real + n.imaginary*n.imaginary)
}

// Exp returns the result of raising Euler's number (e) to the power of the complex number.
func (n Number) Exp() Number {
	magnitude := Number{math.Exp(n.real), 0}
	rotationalFactor := Number{math.Cos(n.imaginary), math.Sin(n.imaginary)}

	return magnitude.Multiply(rotationalFactor)
}
